### ISSUE TITLE
Interpretation Sensitivity and normalized rank of a perfect model

### ISSUE THREAD

**Issue by c1au6i0** [evalclass/precrec#10]
I have some problem interpreting the following graphs that plot Sensitivity vs Normalized Rank of a perfect model.

```
library(precrec)
p <- rbinom(100, 1, 0.5) # same vector for predictions and observations
prc <- evalmod(scores = p, labels = p, mode="basic")

autoplot(prc, c("Specificity", "Sensitivity"))
```
![000009](https://user-images.githubusercontent.com/47305205/54477280-dac17700-47dc-11e9-8191-8b0b5d35ff69.png)

I would expect that a perfect model would generate values of Specificity = Sensitivity = 1 for all the retrieved ranked documents and thus, a line with slope 0 and intercept 1. I am clearly missing something and/or misinterpreting the x axis label. Any hint?

Thanks

**Comment by takayasaito** [comment]
Yes, both specificity and sensitivity should be 1 for all the ranks as you mentioned. It seems like the calculation of normalised ranks is incorrect in this case.

**Comment by takayasaito** [comment]
The current version of precrec does not provide a clear and intuitive calculation method for ties when `mode="basic"` is specified. For example, `rbinom(100, 1, 0.5)` produces many tied scores that are either 0 or 1. The following 3 plots were created by precrec for a dataset generated by `rbinom(100, 1, 0.5)`. 

![image1](https://user-images.githubusercontent.com/10648212/55045394-ecf1af80-503d-11e9-98fc-7ebd6cb32bad.png)

The red dots in the specificity and the sensitivity plots correspond to the red line in the ROC plot. Also, the red dots indicate that specificity stays at 1 while sensitivity changes from 0 to 1. This approach is not incorrect, and it is handy to calculate interpolated ROC and precision-recall curves. Nonetheless, the following plots can be more intuitive when there are tied scores and `mode="basic"` is used. 

![image2](https://user-images.githubusercontent.com/10648212/55045446-18749a00-503e-11e9-85d3-76298e887cf1.png)

The plots above were generated in a way that the same ties have the same sensitivity and specificity. I need to look into whether or not this approach makes sense and how to implement it if so.

**Comment by c1au6i0** [comment]
Thank you for the reply. I think that I understand what you mean and now I get it. The ROC curve really helped.

**Comment by takayasaito** [comment]
Fixed in 0.22.0.

`evalmod(mode = "basic")` now takes a `basic_ties` argument that says what the metrics report at the cutoffs inside a run of tied scores. A cutoff inside such a run separates instances that share a score, so no threshold actually produces it, and the package had only one way of filling it in.

The default `"split"` spreads the run's true and false positives evenly over its cutoffs. That is what the plots above show, and it is the interpolation the ROC and precision-recall curves need, so it stays the default. `"hold"` gives every cutoff of a run the counts it has once the whole run is taken, so tied instances share one value of every metric:

```r
library(precrec)

set.seed(42)
p <- rbinom(100, 1, 0.5)

autoplot(
  evalmod(scores = p, labels = p, mode = "basic", basic_ties = "hold"),
  c("specificity", "sensitivity")
)
```

Sensitivity is 1 from the first rank on, and each metric becomes a step function that changes only where the score does - the second set of plots in the comment above.

One part of the original expectation is not reachable under any tie treatment: specificity still falls to 0 at the right edge. At normalized rank 1 every instance is predicted positive, so there are no true negatives left however good the classifier is.

The two settings agree whenever the scores are all distinct.

CRAN is still at 0.14.5, so this needs the GitHub version for now:

```r
remotes::install_github("evalclass/precrec")
```

The x axis and tied scores are written up at https://evalclass.github.io/precrec/articles/plots-basic-metrics.html
